File: library/builtins/traits/literal.rs
use crate::*;
use wipple::*;

macro_rules! define {
    ($name:ident) => {
        #[derive(TypeInfo, Debug, Clone)]
        pub struct $name {
            pub value: Value,
            pub location: Option<SourceLocation>,
        }

        impl Primitive for $name {}

        impl $name {
            pub fn new(value: Value) -> Self {
                $name::new_located(value, None)
            }

            pub fn new_located(value: Value, location: Option<SourceLocation>) -> Self {
                $name { value, location }
            }
        }
    };
}

define!(Literal);
define!(Escaped);

stored_closure!(pub struct InterpolateFn(bool, &Env, &Stack) -> Result<Value>);

impl Primitive for InterpolateFn {}

#[ext(pub, name = ValueInterpolateExt)]
impl Value {
    // FIXME: Split into two different traits
    fn interpolate(&self, direct: bool, env: &Env, stack: &Stack) -> Result<Cow<Value>> {
        match self.get_if_present::<InterpolateFn>(env, stack)? {
            Some(interpolate) => interpolate(direct, env, stack).map(Cow::Owned),
            None if direct => self.evaluate(env, stack),
            None => Ok(Cow::Borrowed(self)),
        }
    }
}

pub(crate) fn setup(env: &Env, stack: &Stack) -> Result<()> {
    env.set_variable(stack, "Literal", Value::of(Trait::of::<Literal>()))?;

    // Literal == Evaluate
    env.add_relation_between(stack, |literal: Literal| {
        EvaluateFn::new(move |env, stack| {
            literal
                .value
                .interpolate(false, env, stack)
                .map(Cow::into_owned)
        })
    })?;

    // Literal == Text
    env.add_relation_between_with(stack, {
        let env = env.clone();

        move |literal: Literal, stack| {
            let text = literal.value.format(&env, stack)?;
            Ok(Text::new(format!("'{}", text)))
        }
    })?;

    env.set_variable(
        stack,
        "literal",
        Value::of(Function::new(|value, _, _| Ok(value))),
    )?;

    env.set_variable(stack, "Escaped", Value::of(Trait::of::<Escaped>()))?;

    // Escaped == Evaluate
    env.add_relation_between(stack, |escaped: Escaped| {
        EvaluateFn::new(move |env, stack| {
            escaped
                .value
                .interpolate(true, env, stack)
                .map(Cow::into_owned)
        })
    })?;

    // Escaped == Interpolate
    env.add_relation_between(stack, |escaped: Escaped| {
        InterpolateFn::new(move |_, env, stack| {
            escaped.value.evaluate(env, stack).map(Cow::into_owned)
        })
    })?;

    // Escaped == Text
    env.add_relation_between_with(stack, {
        let env = env.clone();

        move |escaped: Escaped, stack| {
            let text = escaped.value.format(&env, stack)?;
            Ok(Text::new(format!("\\{}", text)))
        }
    })?;

    env.set_variable(
        stack,
        "Interpolate",
        Value::of(Trait::of::<InterpolateFn>()),
    )?;

    Ok(())
}

File: Library/Sources/Traits/Literal.swift
import Wipple

public struct Literal: RawRepresentable, Primitive {
    public var rawValue: Value

    public init(rawValue: Value) {
        self.rawValue = rawValue
    }
}

public struct Escaped: RawRepresentable, Primitive {
    public var rawValue: Value

    public init(rawValue: Value) {
        self.rawValue = rawValue
    }
}

// FIXME: Split into two different traits
public struct Interpolate: Primitive {
    public typealias RawValue = (_ direct: Bool, Env, Stack) throws -> Value

    public let rawValue: RawValue

    public init(_ rawValue: @escaping RawValue) {
        self.rawValue = rawValue
    }

    public func callAsFunction(direct: Bool, _ env: Env, _ stack: Stack) throws -> Value {
        try self.rawValue(direct, env, stack)
    }
}

public extension Value {
    func interpolate(direct: Bool, _ env: Env, _ stack: Stack) throws -> Value {
        if let interpolate = try self.getIfPresent(Interpolate.self, env, stack) {
            return try interpolate(direct: direct, env, stack)
        } else if direct {
            return try self.evaluate(env, stack)
        } else {
            return self
        }
    }
}

internal func setupLiteral(_ env: Env, _ stack: Stack) throws {
    try env.setVariable("Literal", to: Value(Trait(Literal.self)), stack)

    // Literal == Evaluate
    try env.addRelation(from: Literal.self, to: Evaluate.self, stack) { literal in
        Evaluate { env, stack in
            try literal.rawValue.interpolate(direct: false, env, stack)
        }
    }

    // Literal == Text
    try env.addRelation(from: Literal.self, to: Text.self, stack) { literal, _, stack in
        try Text(rawValue: "'\(literal.rawValue.format(env, stack))")
    }

    // literal : 'x -> x (TODO: Write in Wipple code)
    try env.setVariable("literal", to: Value(Function { value, _, _ in value }), stack)

    try env.setVariable("Escaped", to: Value(Trait(Escaped.self)), stack)

    // Escaped == Evaluate
    try env.addRelation(from: Escaped.self, to: Evaluate.self, stack) { escaped in
        Evaluate { env, stack in
            try escaped.rawValue.interpolate(direct: true, env, stack)
        }
    }

    // Escaped == Interpolate
    try env.addRelation(from: Escaped.self, to: Interpolate.self, stack) { escaped in
        Interpolate { _, env, stack in
            try escaped.rawValue.evaluate(env, stack)
        }
    }

    // Escaped == Text
    try env.addRelation(from: Escaped.self, to: Text.self, stack) { escaped, _, stack in
        try Text(rawValue: "\\\(escaped.rawValue.format(env, stack))")
    }

    try env.setVariable("Interpolate", to: Value(Trait(Interpolate.self)), stack)
}
